File: 17/17.pl
#!/usr/bin/perl

#
# Copyright (C) 2023 Joelle Maslak
# All Rights Reserved - See License
#

use JTM::Boilerplate 'script';
no warnings 'recursion';    # We know it does deep recursion!
use experimental 'args_array_with_signatures';

use lib '.';
use Table;

use List::Util qw(min sum);
use Data::Dump;
use Memoize;

MAIN: {
    memoize('get_costs');
    my $t = Table->new();
    $t->read( *ARGV, sub { split // } );

    # Part A

    my $dist = Table->new();
    get_dist( $t, $dist, 1, 4 );

    my $start = Coord->new( row => 0, col => 0 );
    my $cost  = Table->new();
    get_costs( $dist, $cost, $start, 'H', 0 );
    get_costs( $dist, $cost, $start, 'V', 0 );

    my $min = min values $cost->get_xy( $cost->row_count() - 1, $cost->col_count() - 1 )->%*;
    say "Heat loss part A: $min";

    # Part B

    $dist = Table->new();
    get_dist( $t, $dist, 4, 10 );

    $start = Coord->new( row => 0, col => 0 );
    $cost  = Table->new();
    get_costs( $dist, $cost, $start, 'H', 0 );
    get_costs( $dist, $cost, $start, 'V', 0 );

    $min = min values $cost->get_xy( $cost->row_count() - 1, $cost->col_count() - 1 )->%*;
    say "Heat loss part B: $min";
}

sub get_dist ( $t, $dist, $mindist, $maxdist ) {
    my $maxi = $t->row_count();
    my $maxj = $t->col_count();

    my @c;
    for ( my $i = 0; $i < $maxi; $i++ ) {
        $c[$i] = [];
        for ( my $j = 0; $j < $maxj; $j++ ) {
            $c[$i]->[$j] = $t->get_xy( $i, $j );
        }
    }
    for ( my $i = 0; $i < $maxi; $i++ ) {
        for ( my $j = 0; $j < $maxj; $j++ ) {
            my %edges = (
                h => [],
                v => [],
            );
            my $running_up = 0;
            my $running_dn = 0;
            for ( my $di = 1; $di <= $maxdist; $di++ ) {
                if ( $i + $di < $maxi ) {
                    $running_up += $c[ $i + $di ]->[$j];
                    if ( $di >= $mindist ) { push $edges{v}->@*, [ $i + $di, $j, $running_up ] }
                }
                if ( $i - $di >= 0 ) {
                    $running_dn += $c[ $i - $di ]->[$j];
                    if ( $di >= $mindist ) { push $edges{v}->@*, [ $i - $di, $j, $running_dn ] }
                }
            }
            $running_up = 0;
            $running_dn = 0;
            for ( my $dj = 1; $dj <= $maxdist; $dj++ ) {
                if ( $j + $dj < $maxj ) {
                    $running_up += $c[$i]->[ $j + $dj ];
                    if ( $dj >= $mindist ) { push $edges{h}->@*, [ $i, $j + $dj, $running_up ]; }
                }
                if ( $j - $dj >= 0 ) {
                    $running_dn += $c[$i]->[ $j - $dj ];
                    if ( $dj >= $mindist ) { push $edges{h}->@*, [ $i, $j - $dj, $running_dn ]; }
                }
            }
            $dist->put_xy( $i, $j, \%edges );
        }
    }
    return;
}

sub get_costs ( $t, $cost, $start, $dir, $init_cost ) {
    # Cost structure is:
    # { R1 => cost, R2 => cost }
    $cost->put( $start, undef );
    my @stack;
    push @stack, [ $start->row(), $start->col(), $dir, $init_cost ];

    while ( my $top = shift @stack ) {
        my ( $start_x, $start_y, $dir, $costval ) = @$top;
        my $me = $cost->get_xy( $start_x, $start_y );
        if ( !defined($me) ) {
            $cost->put_xy( $start_x, $start_y, {} );
        }
        if ( ( !exists( $me->{$dir} ) ) or $costval < $me->{$dir} ) {
            $me->{$dir} = $costval;
        } else {
            next;
        }

        my @nodes;
        my $newdir;
        if ( $dir eq 'V' ) {
            @nodes  = $t->get_xy( $start_x, $start_y )->{h}->@*;
            $newdir = 'H';
        }
        if ( $dir eq 'H' ) {
            @nodes  = $t->get_xy( $start_x, $start_y )->{v}->@*;
            $newdir = 'V';
        }

        for my $node (@nodes) {
            my ( $i, $j, $edgecost ) = @$node;
            push @stack, [ $i, $j, $newdir, $edgecost + $costval ];
        }
    }
    return;
}

File: 17/17.go
package main

import (
	"fmt"
	"math"
	"os"
	"runtime/pprof"
)

type Direction int

const (
	H Direction = iota
	V
)

type Edge struct {
	C    Coord
	Cost int
}

type Edges struct {
	H []Edge
	V []Edge
}

type Costs struct {
	H int
	V int
}

type StackEntry struct {
	C    Coord
	Dir  Direction
	Cost int
}

func main() {
	f, _ := os.Create("prof.prof")
	pprof.StartCPUProfile(f)
	defer pprof.StopCPUProfile()

	var t Table[int]
	t.Read(os.Stdin, func(s string) []int {
		ret := make([]int, len(s), len(s))
		for i, r := range s {
			ret[i] = int(r - '0')
		}
		return ret
	})

	fmt.Printf("Heat loss part A: %d\n", findMinCost(t, 1, 3))
	fmt.Printf("Heat loss part B: %d\n", findMinCost(t, 4, 10))
}

func findMinCost(t Table[int], minDist int, maxDist int) int {
	edges := getEdges(t, minDist, maxDist)
	costs := getCosts(edges)

	nodeCosts := costs.GetXY(t.RowCount()-1, t.ColCount()-1)
	if nodeCosts.H < nodeCosts.V {
		return nodeCosts.H
	}
	return nodeCosts.V
}

func getEdges(t Table[int], minDist int, maxDist int) Table[Edges] {
	maxi := t.RowCount()
	maxj := t.ColCount()

	var edges Table[Edges]

	for i, row := range t.GetRows() {
		for j := range row.GetCols() {
			var e Edges

			// Vertical edges
			sumUp := 0
			sumDn := 0
			for di := 1; di <= maxDist; di++ {
				if i+di < maxi {
					sumUp += t.GetXY(i+di, j)
					c := Coord{row: i + di, col: j}
					if di >= minDist {
						e.V = append(e.V, Edge{C: c, Cost: sumUp})
					}
				}
				if i-di >= 0 {
					sumDn += t.GetXY(i-di, j)
					c := Coord{row: i - di, col: j}
					if di >= minDist {
						e.V = append(e.V, Edge{C: c, Cost: sumDn})
					}
				}
			}

			// Horizontal edges
			sumUp = 0
			sumDn = 0
			for dj := 1; dj <= maxDist; dj++ {
				if j+dj < maxj {
					sumUp += t.GetXY(i, j+dj)
					c := Coord{row: i, col: j + dj}
					if dj >= minDist {
						e.H = append(e.H, Edge{C: c, Cost: sumUp})
					}
				}
				if j-dj >= 0 {
					sumDn += t.GetXY(i, j-dj)
					c := Coord{row: i, col: j - dj}
					if dj >= minDist {
						e.H = append(e.H, Edge{C: c, Cost: sumDn})
					}
				}
			}

			// Write out to table
			edges.PutXY(i, j, e)
		}
	}

	return edges
}

func getCosts(edges Table[Edges]) Table[Costs] {
	var costs Table[Costs]
	for i := edges.RowCount() - 1; i >= 0; i-- { // Backwards because more efficient
		for j := edges.ColCount() - 1; j >= 0; j-- {
			costs.PutXY(i, j, Costs{H: math.MaxInt, V: math.MaxInt})
		}
	}

	stack := make([]StackEntry, 2, 2)
	stack = append(stack, StackEntry{Dir: H}, StackEntry{Dir: V})
	for len(stack) > 0 {
		top := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		me := costs.Get(top.C)
		if top.Dir == V && top.Cost < me.V {
			me.V = top.Cost
		} else if top.Dir == H && top.Cost < me.H {
			me.H = top.Cost
		} else {
			// Not cheaper
			continue
		}
		costs.Put(top.C, me)

		// We change direction for the next visits!
		if top.Dir == V {
			for _, visit := range edges.Get(top.C).H {
				stack = append(stack, StackEntry{C: visit.C, Cost: visit.Cost + top.Cost, Dir: H})
			}
		} else {
			for _, visit := range edges.Get(top.C).V {
				stack = append(stack, StackEntry{C: visit.C, Cost: visit.Cost + top.Cost, Dir: V})
			}
		}
	}

	return costs
}
